
""" rules for scraps counting """

rule assign_sites_read2:
  input:
    "{results}/{sample}/{sample}_{read}_Aligned.sortedByCoord.out.bam"
  output:
    bam = "{results}/counts/{sample}_{read}.bam",
    bai = "{results}/counts/{sample}_{read}.bam.bai"
  params:
    job_name = "assign_sites_read2",
    memory = "select[mem>8] rusage[mem=8]", # LSF format; change as needed
    out = "{results}/counts/{sample}_{read}_assigned",
    out_bam = "{results}/counts/{sample}_{read}_Aligned.sortedByCoord.out.bam.featureCounts.bam"
  log:
    "logs/assign_sites_{sample}_{read}.txt"
  threads:
    12
  shell:
    """
    if echo {POLYA_SITES} | tr '[:upper:]' '[:lower:]' | grep -q -e "\.saf$" -e "\.saf.gz$"; then
      polyaformat='SAF'; else
      polyaformat='GTF';
    fi

    featureCounts \
    -s 1 -Q 10 -O \
    --read2pos 3 \
    -o {params.out} \
    -F $polyaformat \
    -a {POLYA_SITES} \
    -R BAM \
    -T {threads} \
    {input} \
    2> {log}
    
    samtools sort \
    {params.out_bam} \
    -o {output.bam}
    
    samtools index {output.bam}
    """

rule filter:
  input:
    "{results}/counts/{sample}_R2.bam"
  output:
    "{results}/counts/{sample}_R2_filter.done"
  params:
    job_name = "filter",
    temp = "{results}/counts/{sample}_R2_filter.bam",
    memory = "select[mem>8] rusage[mem=8]" # LSF format; change as needed
  log:
    "logs/filter_{sample}_R2.txt"
  threads:
    1
  shell:
    """
    samtools view -h {input} | grep -v 'CB:Z:-\|UB:Z:-' | samtools view -b - > {params.temp}
    samtools index {params.temp}
    python3 ~/projects/scraps_comp/code/scraps/inst/scripts/filter_bam.py -i {params.temp} -o {input}
    samtools index {input}
    touch {output}
    """

rule count:
  input:
    "{results}/counts/{sample}_{read}.bam",
    "{results}/counts/{sample}_R2_filter.done"
  output:
    "{results}/counts/{sample}_{read}_counts.tsv.gz"
  params:
    job_name = "count",
    memory = "select[mem>8] rusage[mem=8]" # LSF format; change as needed
  log:
    "logs/count_{sample}_{read}.txt"
  threads:
    1
  shell:
    """
    umi_tools count \
      --per-gene \
      --gene-tag=XT \
      --assigned-status-tag=XS \
      --per-cell \
      --extract-umi-method=tag \
      --umi-tag=UB \
      --cell-tag=CB \
      -I {params.temp} \
      -S {output} \
      > {log}
    """
